src/axom/core/FlatMap.hpp

```diff
@@ -0,0 +1,818 @@
+   */
+  bool empty() const { return m_size == 0; }
+
+  /*!
+   * \brief Returns the number of entries stored in the FlatMap.
+   */
+  IndexType size() const { return m_size; }
+
+  /*!
+   * \brief Try to find an entry with a given key.
+   *
+   * \param [in] key the key to search for
+   *
+   * \return An iterator pointing to the corresponding key-value pair, or end()
+   *  if the key wasn't found.
+   */
+  /// @{
+  iterator find(const KeyType& key);
+  const_iterator find(const KeyType& key) const;
+  /// @}
+
+  /*!
+   * \brief Try to find an entry with a given key.
+   *
+   * \param [in] key the key to search for
+   *
+   * \return A reference to the corresponding value.
+   * \throw std::out_of_range if the key is not found.
+   */
+  /// @{
+  ValueType& at(const KeyType& key)
+  {
+    auto it = this->find(key);
+    if(it == end())
+    {
+      throw std::out_of_range {"axom::FlatMap::find(): key not found"};
+    }
+    return it->second;
+  }
+  const ValueType& at(const KeyType& key) const
+  {
+    auto it = this->find(key);
+    if(it == end())
+    {
+      throw std::out_of_range {"axom::FlatMap::find(): key not found"};
+    }
+    return it->second;
+  }
+  /// @}
+
+  /*!
+   * \brief Find an entry with a given key.
+   *
+   *  If a corresponding value does not exist, a default value for the value
+   *  type will be inserted for the given key.
+   *
+   * \param [in] key the key to search for
+   *
+   * \return A reference to the corresponding value.
+   *
+   * \pre ValueType is default-constructible
+   */
+  /// @{
+  ValueType& operator[](const KeyType& key)
+  {
+    static_assert(std::is_default_constructible<ValueType>::value,
+                  "Cannot use axom::FlatMap::operator[] when value type is not "
+                  "default-constructible.");
+    return this->try_emplace(key).first->second;
+  }
+  const ValueType& operator[](const KeyType& key) const
+  {
+    static_assert(std::is_default_constructible<ValueType>::value,
+                  "Cannot use axom::FlatMap::operator[] when value type is not "
+                  "default-constructible.");
+    return this->try_emplace(key).first->second;
+  }
+  /// @}
+
+  /*!
+   * \brief Return the number of entries matching a given key.
+   *
+   *  This method will always return 0 or 1.
+   *
+   * \param [in] key the key to search for
+   */
+  IndexType count(const KeyType& key) const
+  {
+    return contains(key) ? IndexType {1} : IndexType {0};
+  }
+
+  /*!
+   * \brief Return true if the FlatMap contains a key, false otherwise.
+   *
+   * \param [in] key the key to search for
+   */
+  bool contains(const KeyType& key) const { return (find(key) != end()); }
+
+  /*!
+   * \brief Erases all elements from the FlatMap.
+   */
+  void clear()
+  {
+    // Destroy all elements.
+    IndexType index = this->nextValidIndex(m_metadata, NO_MATCH);
+    while(index < bucket_count())
+    {
+      m_buckets[index].get().~KeyValuePair();
```

Comment:
What would happen when the key type's destructor is called again when `m_buckets` goes out of scope?  Do we need to state that the key type should allow multiple call to the destructor?

Reply:
Below this, the metadata also gets reset, which marks all the destroyed buckets as "empty." This should prevent the objects from getting double-destroyed.